### DISCUSSION TITLE
How can I get to know whether my program is running in drm?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#14879] [Q&A] [answered]
I want to know whether my program is running in drm. Previously I implement it like this:
```
using Avalonia;
using System;
using System.Globalization;
using System.Linq;
using System.Threading;

namespace Test
{
    internal class Program
    {
        // Initialization code. Don't use any Avalonia, third-party APIs or any
        // SynchronizationContext-reliant code before AppMain is called: things aren't initialized
        // yet and stuff might break.
        [STAThread]
        public static int Main(string[] args)
        {
            var builder = BuildAvaloniaApp();
            double GetScaling()
            {
                var idx = Array.IndexOf(args, "--scaling");
                if (idx != 0 && args.Length > idx + 1 &&
                    double.TryParse(args[idx + 1], NumberStyles.Any, CultureInfo.InvariantCulture, out var scaling))
                    return scaling;
                return 1;
            }
            if (args.Contains("--drm"))
            {
                SilenceConsole();
                DefaultProperty.IsDRM = true;
                return builder.StartLinuxDrm(args: args, card: "/dev/dri/card0", scaling: GetScaling());
            }

            return builder.StartWithClassicDesktopLifetime(args);
        }

        private static void SilenceConsole()
        {
            new Thread(() =>
            {
                Console.CursorVisible = false;
                while (true)
                    Console.ReadKey(true);
            })
            { IsBackground = true }.Start();
        }

        // Avalonia configuration, don't remove; also used by visual designer.
        public static AppBuilder BuildAvaloniaApp()
            => AppBuilder.Configure<App>()
                .UsePlatformDetect()
                .WithInterFont()
                .LogToTrace();
    }
}
```

As you see, I used a static Boolean variable DefaultProperty.IsDRM to mark it.

However, when the program running in raspberry pi, it reports this error:
Unhandled exception. System.PlatformNotSupportedException: Operation is not supported on this platform.
   at Avalonia.Threading.Dispatcher.MainLoop(CancellationToken cancellationToken)
   at Avalonia.Controls.DesktopApplicationExtensions.Run(Application app, CancellationToken token)
   at LinuxFramebufferPlatformExtensions.StartLinuxDirect(AppBuilder builder, String[] args, IOutputBackend outputBackend, IInputBackend inputBackend)
   at LinuxFramebufferPlatformExtensions.StartLinuxDrm(AppBuilder builder, String[] args, String card, Double scaling, IInputBackend inputBackend)
   at Test.Program.Main(String[] args) in D:\Project\Test\Test\Program.cs:line 31

If I delete the code:
`                DefaultProperty.IsDRM = true;`

All works well.

I didn't quite understand why setting variables in Program.cs would cause such an error. Later, I tried to set variables in Program.cs on the Windows platform, and sure enough, the same error would be reported. It seems that variables cannot be set in Program.cs.

On the one hand I want to understand why the variable cannot be set in Program.cs, on the other hand I want to know if there is any other way to get whether the program is running under DRM.

Thank you.

**Comment by thevortexcloud** [comment]
I think you can check the lifetime?

https://github.com/AvaloniaUI/Avalonia/blob/9646164708adb1a90f54b8a461b543ad82bef980/src/Linux/Avalonia.LinuxFramebuffer/LinuxFramebufferPlatform.cs#L91

 I don't think there is a way to check for DRM mode specifically though by doing that though. But it should mean at least DRM or Framebuffer mode is active.

**Reply by kyuranger** [reply]
I tried to check it like this, but unfortunately LinuxFramebufferLifetime is a private type and I can't use it.
```
if (Application.Current.ApplicationLifetime is LinuxFramebufferLifetime)
{ 
}
```

**Reply by thevortexcloud** [reply, marked as answer]
You _should_ be able to use a bit of reflection to check the type. Private modifiers are just a compile time restriction, as such it does not really mean anything if you can use reflection to check it at runtime. Obviously doing that is a hack and may break in future, but it will work. Something like this (I have not tested it) is probably the simplest way:

```csharp
private bool IsDRM() {
     return ApplicationLifetime.GetType().Name == "LinuxFramebufferLifetime"
}
```

If someone suggests something better I would strongly suggest doing that instead of this.

**Reply by kyuranger** [reply]
Sorry I didn't get a chance to test this until I was at work today. The test results on the Raspberry Pi are the same as what you provided, proving that your way is feasible. Thank you very much.

**Reply by thevortexcloud** [reply]
You are welcome. I do wish I knew a better way. I am not happy with how big of a hack this is. 

That said I am still confused as to why setting a variable in your program class causes things to break. I have set variables in mine before without issue.

**Comment by jp2masa** [comment]
> If I delete the code:
> `                 DefaultProperty.IsDRM = true;`
>
> All works well.

You probably have some hidden dependency there. Can you provide a minimal repro?

**Reply by kyuranger** [reply]
No, it's just a plain Boolean:
` public static Boolean IsDRM = false;`

**Reply by kyuranger** [reply]
Regarding the error reported by my code, I accidentally saw a default comment in Program.cs: 
// Initialization code. Don't use any Avalonia, third-party APIs or any
         // SynchronizationContext-reliant code before AppMain is called: things aren't initialized
         // yet and stuff might break.
I think this may be the reason

**Reply by jp2masa** [reply]
You're not using anything that depends on Avalonia or its synchronization context, so this is very weird.
